File: BinarySearch/cpp/binary_search_to_find_x_in_sorted_array.cpp
//
// Created by santosh on 12/21/25.
//

#include <bits/stdc++.h>
using namespace std;

class Solution
{
    public:
    int search(vector<int> &nums, int target)
    {
        int l = 0, r = nums.size() - 1;
        while (l <= r)
        {
            int mid = l+(r-l)/2;
            if (nums[mid] == target)
            {
                return mid;
            }
            if (nums[mid] < target)
            {
                l = mid + 1;
            }
            else
            {
                r = mid - 1;
            }
        }
        return -1;
    }
};

int main()
{
    vector<int> arr{1,2,3,4,5,6,7,8,8,8,9,10};
    Solution s;
    int result = s.search(arr,8);
    cout << result << endl;

}

File: BinarySearch/python/binary_search_to_find_x_in_sorted_array.py
# Binary Search to find x in a sorted array (LeetCode 704)
# Created by santosh

class Solution:
    def search(self, nums: list[int], target: int) -> int:
        """
        Standard binary search.
        Time: O(log n), Space: O(1)
        """
        l, r = 0, len(nums) - 1
        while l <= r:
            mid = l + (r - l) // 2
            if nums[mid] == target:
                return mid
            elif nums[mid] < target:
                l = mid + 1
            else:
                r = mid - 1
        return -1


if __name__ == "__main__":
    arr = [1, 2, 3, 4, 5, 6, 7, 8, 8, 8, 9, 10]
    s = Solution()
    result = s.search(arr, 8)
    print(result)
